apps/web/src/app/pages/trade/bridge/_xcall/EvmWalletXService.ts

```diff
@@ -23,15 +26,39 @@ export class EvmWalletXService extends EvmPublicXService implements IWalletXServ
 
   async approve(token, owner, spender, currencyAmountToApprove) {}
 
-  async executeTransfer(xTransactionInput: XTransactionInput) {
-    const { direction, inputAmount, recipient, account, xCallFee } = xTransactionInput;
+  async executeTransaction(xTransactionInput: XTransactionInput) {
+    const { type, direction, inputAmount, recipient, account, xCallFee, executionTrade, slippageTolerance } =
+      xTransactionInput;
+
+    if (!this.walletClient) {
+      throw new Error('Wallet client not found');
+    }
+
+    const receiver = `${direction.to}/${recipient}`;
+    const tokenAddress = inputAmount.wrapped.currency.address;
+    const amount = BigInt(inputAmount.quotient.toString());
+    const destination = `${ICON_XCALL_NETWORK_ID}/${bnJs.Router.address}`;
 
-    if (this.walletClient) {
-      const receiver = `${direction.to}/${recipient}`;
-      const tokenAddress = inputAmount.wrapped.currency.address;
-      const amount = BigInt(inputAmount.quotient.toString());
-      const destination = `${ICON_XCALL_NETWORK_ID}/${bnJs.Router.address}`;
-      const data = toHex(
+    let data: Address;
+    if (type === XTransactionType.SWAP) {
+      if (!executionTrade || !slippageTolerance) {
+        return;
+      }
+      const minReceived = executionTrade.minimumAmountOut(new Percent(slippageTolerance, 10_000));
+      const rlpEncodedData = Buffer.from(
```

Comment:
create a utility function that calculates rlpEncodedData.
use the function in all XServices.

Reply:
i think, it's already optimized to use utility functions.
can you please give me the declaration of the new utility function?

Reply:
done